pennylane/measurements.py

```diff
@@ -288,6 +291,85 @@ def shape(self, device=None):
             + f"{self.return_type}."
         )
 
+    @functools.lru_cache()
+    def _shape_new(self, device=None):
+        """The expected output shape of the MeasurementProcess.
+
+        Note that the output shape is dependent on the device when:
+
+        * The ``return_type`` is either ``Probability``, ``State`` (from :func:`.state`) or
+          ``Sample``;
+        * The shot vector was defined in the device.
+
+        For example, assuming a device with ``shots=None``, expectation values
+        and variances define ``shape=(,)``, whereas probabilities in the qubit
+        model define ``shape=(2**num_wires)`` where ``num_wires`` is the
+        number of wires the measurement acts on.
+
+        Args:
+            device (.Device): a PennyLane device to use for determining the
+                shape
+
+        Returns:
+            tuple: the output shape
+
+        Raises:
+            QuantumFunctionError: the return type of the measurement process is
+                unrecognized and cannot deduce the numeric type
+        """
+        shape = None
+
+        # First: prepare the shape for return types that do not require a
+        # device
+        if self.return_type in (Expectation, MutualInfo, Variance, VnEntropy):
+            shape = ()
+
+        density_matrix_return = self.return_type == State and self.wires
+
+        if density_matrix_return:
+            dim = 2 ** len(self.wires)
+            shape = (dim, dim)
+
+        # Determine shape if device with shot vector
+        if device is not None and device._shot_vector is not None:
+            shape = self._shot_vector_shape(device, main_shape=shape)
+
+        # If we have a shape, return it here
+        if shape is not None:
+            return shape
+
+        # Then: handle return types that require a device; no shot vector
+        if device is None and self.return_type in (Probability, State, Sample):
```

Comment:
Do you think it would be useful to change that behavior and assume a qubit device (just using self.wires) with a warning message? Then we could have tape.shape without the use of a device.

Reply:
Maybe it would be better. But for this PR I'm focusing on only updating the output shapes without changing any functionality.